ASoC: SOF: file layout profile handling with fallback mechanism (take 3)

ujfalusi opened the pull request:
Hi,

replaces #4309 (it is an iteration on the same idea, but significantly different to confuse GH, I think).

This series aims to add support fro IPC type fallback in case the default firmware file or the topology file is missing  from the filesystem.
If a user requested firmware is loaded (via module parameter) then the IPC type is adjusted to match with the loaded firmware to make sure that we are looking for the matching topology file.

I'm moving all logic to core from sof-acpi/of/pci-dev files, moving also the `sof_machine_*` functions locally.
The change is gradual, first I just move the path/name configuration, then changing the order of initialization to be able to query the topology file earlier, then when all information is available, we can do the file existence checks and fallback handling.

ujfalusi committed:
ASoC: SOF: Add placeholder for platform IPC type and path overrides

Add a struct sof_loadable_file_profile which can be filled by platforms
(sof-acpi-dev.c, sof-of-dev.c and sof-acpi-dev.c) to be able to use common,
generic code to handle path customization.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-acpi-dev: Save the default IPC type and path overrides

Store the default IPC type and the firmware and topology path overrides to
ipc_file_profile_base

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-of-dev: Save the default IPC type and path overrides

Store the default IPC type and the firmware and topology path overrides to
ipc_file_profile_base

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-pci-dev: Save the default IPC type and path overrides

Store the default IPC type and the overrides to ipc_file_profile_base

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: core: Implement firmware, topology path setup in core

Use the information stored in ipc_file_profile_base by platforms to
construct the paths, filenames that are going to be used to load the
firmware and topology files.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-acpi-dev: Rely on core to create the file paths

The core is now using the information from ipc_file_profile_base to create
the paths for the loadable files, no need to set it in here anymore.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-of-dev: Rely on core to create the file paths

The core is now using the information from ipc_file_profile_base to create
the paths for the loadable files, no need to set it in here anymore.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: sof-pci-dev: Rely on core to create the file paths

The core is now using the information from ipc_file_profile_base to create
the paths for the loadable files, no need to set it in here anymore.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Move sof_of_machine_select() to sof-of-dev.c from sof-audio.c

Move the sof_of_machine_select() function to sof-of-dev.c file and provide
an inline stub in case of non OF builds.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Move sof_machine_* functions from sof-audio.c to core.c

Relocate the machine handling functions from sof-audio.c to core.c to
maintain code separation.

While doing the move, drop the redundant
IS_ERR_OR_NULL(plat_data->pdev_mach) check from
sof_machine_unregister()

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: core: Add helper for initialization of paths, ops

Add sof_init_environment() as a helper function to contain path and ops
initialization.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: core: Implement IPC version fallback if firmware files are missing

If a firmware file is missing for the selected IPC type then try to switch
to other supported IPC type and check if that one can be used instead.

If for example a platform is changed to IPC4 as default version but the
given machine does not yet have the needed topology file created then we
will fall back to IPC3 which should have all the needed files.

Relocate the sof_init_environment() to be done at a later phase, in
sof_probe_continue().

This will only have changes in behavior if
CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE is enabled (Intel HDA platforms) by
not failing the module probe, but it is not going to be different case
compared to for example failed firmware booting or topology loading error.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

paulstelian97 commented:
I assume this is only the infra change. Can you show an example of how this new infra is to be used? (I do like how this looks but an example is still useful).

ujfalusi commented:
@paulstelian97, sure!
```diff
diff --git a/sound/soc/sof/intel/pci-tgl.c b/sound/soc/sof/intel/pci-tgl.c
index ca37ff1bbd2a..24dc3efce1ed 100644
--- a/sound/soc/sof/intel/pci-tgl.c
+++ b/sound/soc/sof/intel/pci-tgl.c
@@ -29,7 +29,7 @@ static const struct sof_dev_desc tgl_desc = {
 	.irqindex_host_ipc      = -1,
 	.chip_info = &tgl_chip_info,
 	.ipc_supported_mask	= BIT(SOF_IPC) | BIT(SOF_INTEL_IPC4),
-	.ipc_default		= SOF_IPC,
+	.ipc_default		= SOF_INTEL_IPC4,
 	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
 	.default_fw_path = {
 		[SOF_IPC] = "intel/sof",
```
1. IPC4 firmware and topology file is present:
```
snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw path: community
snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw_lib path: community
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3: Loadable file profile for ipc type 1:
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file:     intel/avs/tgl/community/dsp_basefw.bin
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware lib path: intel/avs-lib/tgl/community
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file:     intel/avs-tplg/sof-hda-generic.tplg
```
2. IPC4 firmware or topology file is missing (fallback to IPC3):
```
snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw path: community
snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw_lib path: community
sof-audio-pci-intel-tgl 0000:00:1f.3: Using fallback IPC type 0 (requested type was 1)
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3: Loadable file profile for ipc type 0:
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file:     intel/sof/community/sof-tgl.ri
snd_sof:sof_print_profile_info: sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file:     intel/sof-tplg/sof-hda-generic.tplg
```
3. Neither IPC4 or IPC3 files are present:
```
[18901.573746] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw path: community
[18901.573748] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw_lib path: community
[18901.574089] sof-audio-pci-intel-tgl 0000:00:1f.3: SOF firmware and/or topology file not found.
[18901.574093] sof-audio-pci-intel-tgl 0000:00:1f.3: Available default profiles
[18901.574095] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 0 (Fallback):
[18901.574097] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/sof/community/sof-tgl.ri
[18901.574099] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/sof-tplg/sof-hda-generic.tplg
[18901.574101] sof-audio-pci-intel-tgl 0000:00:1f.3: Available default profiles
[18901.574102] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 1 (Requested):
[18901.574104] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/avs/tgl/community/dsp_basefw.bin
[18901.574105] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/avs-tplg/sof-hda-generic.tplg
[18901.574106] sof-audio-pci-intel-tgl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
[18901.574107] sof-audio-pci-intel-tgl 0000:00:1f.3: Optionally it can be manually downloaded from:
[18901.574108] sof-audio-pci-intel-tgl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
```
Before this series we would fail if a file is not present at the time when we try to load it later in the boot sequence, but with this change we can detect it earlier and do the fallback.
This is a feature that would help distributions (and us) cope with migrating to a new IPC type as painfully as it can be done. It is not good if users need to modify files in `/etc/` to get their audio back.

ujfalusi commented:
Hrm, why it is printing it twice twice in case of neither is there??? Oops.

ujfalusi commented:
OK, fixed it and reworded to: "Supported default profiles"

paulstelian97 commented:
I definitely like this, what happens if a non-default type was requested via module parameter and it wasn't found? (or even if the default was requested by parameter). Would the fallback still happen or would it explicitly be disabled?

I definitely have some interest in this as I actually have the migrating to IPC4 task (for i.MX platforms) in my backlog (no target version yet), this change definitely should help smooth out the transition.

ujfalusi commented:
It works in the same way, the module parameter takes the place of the default ipc type and the same course of events will happen.
We will fail if the requested type is not supported (as before), but we will fallback if any of the needed files are missing.

ujfalusi commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -92,11 +194,65 @@ sof_file_profile_for_ipc_type(struct snd_sof_dev *sdev,
+		else
+			marker = "Fallback";
+
+		dev_info(dev, "Available default profiles\n");
```
this should be outside of the loop..

ujfalusi committed:
ASoC: SOF: Intel: Do not use resource managed allocation for ipc4_data

Manage the ipc4_data allocation in code instead of devm since the ops_init
might be called more than once due to IPC type fallback.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- Add patches for Intel platforms to manage the ipc4_data instead of devm resource managing to be able to free it when changing IPC type due to fallback
- Use dev_dbg() print to notify developers about the missing file (otherwise the fallback is silent to not alarm users)
- Clean up of prints and smaller cleanups to make the code easier to follow and understand.

mengdonglin commented:
@ujfalusi Would this PR be merged before https://github.com/thesofproject/linux/pull/4301 that introduces generic IPC type identifiers?

ujfalusi commented:
> @ujfalusi Would this PR be merged before #4301 that introduces generic IPC type identifiers?

I would prefer that way to make it easier to backport if needed.

marc-hb commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -0,0 +1,311 @@
+	if (profile->fw_lib_path)
+		dev_dbg(dev, " Firmware lib path: %s\n", profile->fw_lib_path);
+
+	dev_dbg(dev, " Topology file:     %s/%s\n", profile->tplg_path,
```
Can you please increase the log levels when logging the loaded files? To be honest this is something `request_firmware()` itself should do (imagine that: a "bootloader" that does not even log what it boots) but I guess this is unlikely to happen any time soon.

Some more context in
- #3867

ujfalusi replied:
I don't think normal users care about this at all and just going to make the kernel log noisy for no benefit.

Adding a debugfs file would be a better solution if we need it (separate files or one file which prints the paths and names?)

marc-hb replied:
Define "normal users". Anyone staring at kernel logs and _not_ caring what firmware and configuration files are in use is seriously shooting themselves in the foot.

The kernel squeals "TAINTED" as soon as something is slightly out of the ordinary but somehow loading unknown files from random locations (`/lib/firmware/updates/`, 
`/lib/firmware/UTS_RELEASE/`, etc.) does not deserve a word.

And please don't get me started on the (lack of) security for which logging is one of the most basic requirements.

>  make the kernel log noisy for no benefit.

In "normal" usage, files are loaded only ONCE per boot so this would be obviously not noisy at all.

It would be printed more often only when running a test like "kmod-unload-load" but then such logs become even more relevant.

ujfalusi replied:
OK.
Let's make the path, name prints `dev_info()`

There is one thing I'm not happy about is this:
```
dev_dbg(dev, "Loadable file profile for ipc type %d:\n", profile->ipc_type);
```
Should it be simply:
```
dev_info(dev, "Firmware paths/files for ipc type %d:\n", profile->ipc_type);
```
Despite the fact the the topology is not really a firmware and the libraries are under a path, we don't know what will be loaded from there?

ujfalusi commented:
Changes since v2:
- use `dev_info()` for printing the paths/file names
```
[ 6693.796773] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware paths/files for ipc type 1:
[ 6693.796778] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file:     intel/sof-ipc4/tgl/community/sof-tgl.ri
[ 6693.796780] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware lib path: intel/sof-ipc4-lib/tgl/community
[ 6693.796782] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file:     intel/sof-ipc4-tplg/sof-hda-generic.tplg
```

ujfalusi commented:
Changes since v3:
- rebased on topic/sof-dev

ujfalusi commented:
Changes since v4:
- rebased on topic/sof-dev

plbossart commented:
@ujfalusi I've completely lost the plot on this one. 

In what case would it happen that "default firmware file or the topology file is missing from the filesystem" ?

ujfalusi commented:
> @ujfalusi I've completely lost the plot on this one.
> 
> In what case would it happen that "default firmware file or the topology file is missing from the filesystem" ?

@plbossart, we would like to change the default IPC from 3 to 4 for TGL family but it cannot be done as a brute force as the kernel and firmware/topology is coming from different and non synchronized sources.
We are going to provide IPC3 firmware/topologies for TGL family in releases but at the same time we would also like to have IPC4 firmware for the platforms (not all might have the firmware in all releases) and on top of this we are lagging a bit behind on topologies as well, so even if there is a firmware, we might not have the topology.

In order to protect the user form this transition (Linux users, such as Arch/ubuntu/fedora/SUSE/etc) we need to have a mechanism which will fall back to IPC3 when IPC4 is missing a firmware/topology file.
The dmesg will be clear on what happens and the user will have working audio, no need to manually add module parameters.

The logic is really not trying to be clever and invent things up. We just check if the files are there and if they are then we go with that selection, if not then fallback to a tried and working IPC version. Atm we have IPC4 and IPC3, but the code should be able to handle newer IPC versions and cascade down to find a working setup.

It also helps customers/developers where they have IPC3/IPC4 topologies in standard paths but they load different version of firmware by providing custom paths: if the firmware is IPC4 then we are going to look for IPC4 topologies, if it is IPC3 then IPC3 topologies will be requested.

On the way to achieve this (I think) we have a nice cleanup around the file and path handling, one place for all bus type (acpi/of/pci) with the logic in one place, easier to extend, enhance, blame and fix if there is a need.

@kv2019i, am I forgot something?

plbossart commented:
well there's another (simpler) way of achieving this transition: you add a SOF_TGL_IPC4 Kconfig that defaults to 'n' and needs to be selected by distros, and when they do select SOF_TGL_IPC4 then they use the files that are managed by the same distribution.

I don't think it's wise to change the default for everyone and then have a fallback.

ujfalusi commented:
> well there's another (simpler) way of achieving this transition: you add a SOF_TGL_IPC4 Kconfig that defaults to 'n' and needs to be selected by distros, and when they do select SOF_TGL_IPC4 then they use the files that are managed by the same distribution.
> 
> I don't think it's wise to change the default for everyone and then have a fallback.

How can Ubuntu select IPC4 for all of their users? They don't own every single machine to make sure that all of them have topology support and so forth...

plbossart commented:
> > well there's another (simpler) way of achieving this transition: you add a SOF_TGL_IPC4 Kconfig that defaults to 'n' and needs to be selected by distros, and when they do select SOF_TGL_IPC4 then they use the files that are managed by the same distribution.
> > I don't think it's wise to change the default for everyone and then have a fallback.
> 
> How can Ubuntu select IPC4 for all of their users? They don't own every single machine to make sure that all of them have topology support and so forth...

so even the SOF developer team has no idea which devices can actually be supported with this transition, so why even think of making it a virtual default and fallback. We have not moved one inch further than the initial opt-in support, so my take is that it has to stay opt-in until we get our act together and have a clear plan topology-wise. We should not push this upstream, it's half-baked.

plbossart commented:
> > > well there's another (simpler) way of achieving this transition: you add a SOF_TGL_IPC4 Kconfig that defaults to 'n' and needs to be selected by distros, and when they do select SOF_TGL_IPC4 then they use the files that are managed by the same distribution.
> > > I don't think it's wise to change the default for everyone and then have a fallback.
> > 
> > 
> > How can Ubuntu select IPC4 for all of their users? They don't own every single machine to make sure that all of them have topology support and so forth...
> 
> so even the SOF developer team has no idea which devices can actually be supported with this transition, so why even think of making it a virtual default and fallback. We have not moved one inch further than the initial opt-in support, so my take is that it has to stay opt-in until we get our act together and have a clear plan topology-wise. We should not push this upstream, it's half-baked.

Or at minimum add a Kconfig to make sure users are not surprised by a change in default that adds an extra step and new messages.

ujfalusi commented:
@plbossart, this PR will _not_ change any default IPC versions, it will just re-arranges the IPC version handling, firmware file location handling and makes it possible to avoid obviously not usable selections.
The kernel has no means to force a certain sof-firmware package on the filesystem, it is better if the kernel is prepared for such case than just throw it's hands up and leave the user without a working sound card after a kernel update (again, if we ever make the IPC4 'default' - but I would use preferred - for the TGL family).

plbossart commented:
I find it weird that we prepare for a fallback without using that fallback code, it's rather unusual.

ujfalusi commented:
> I find it weird that we prepare for a fallback without using that fallback code, it's rather unusual.

We need to have the fallback support to even think about the possibility of taking advantage of it. I really think that IPC4 brings valuable features for users of existing platforms that worth the trouble to try to make it available for the 'masses'.

Even if we are not going to use actively the fallback support, the cleanup and re-structuring of the code will open up possibilities for further improvements or new features around the IPC type and firmware file path/names.

This is why the emphasis of the PR is on IPC type 'profile' support _with_ fallback support, the fallback made possible by the re-structured code.

I'm not saying that this is not controversial, it is, but imho it brings value.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -512,6 +566,41 @@ int snd_sof_device_shutdown(struct device *dev)
+{
+	struct snd_sof_pdata *plat_data = pdata;
+
+	if (!IS_ERR_OR_NULL(plat_data->pdev_mach))
```
while we are doing this move, we might as well remove the OR_NULL here, I can't recall when it was changed or why w

@andy-shev what do you think?

ashevche replied:
This complete check is redundant. `platform_device_unregister()` is aware of both errornous cases. See 99fef587ff98 ("driver core: platform: Respect return code of platform_device_register_full()").

plbossart commented on include/sound/sof.h:
```diff
@@ -57,6 +57,17 @@ enum sof_ipc_type {
+	const char *fw_path;
+	const char *fw_name;
+	const char *fw_lib_path;
+	const char *fw_path_postfix;
```
It's Monday morning and I don't recall what this is? is this a new property or is my memory failing me?

ujfalusi replied:
We are adding a postfix to the firmware/library path in case of community key signed binaries.
This is to hold that postfix, but in a generic way, not tied to the word 'community'

plbossart replied:
ah ok, probably a comment would help to explain that the same postfix will be used for path and lib.

Note that there are already exceptions to the rule, the fw and lib may be signed with different keys. Assuming that both share the same keys and use the same prefix as a consequence will not work in real life for Windows devices.

ujfalusi replied:
> Note that there are already exceptions to the rule, the fw and lib may be signed with different keys. Assuming that both share the same keys and use the same prefix as a consequence will not work in real life for Windows devices.

:astonished:

plbossart replied:
This hasn't changed, we need support for separate keys for fw and libs. Not my decision, it's how the hardware works and how Windows distributed firmware since SKL.

plbossart commented on sound/soc/sof/sof-pci-dev.c:
```diff
@@ -319,6 +320,22 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
+	if (dmi_check_system(community_key_platforms) &&
+	    sof_dmi_use_community_key) {
+		path_override->fw_path_postfix = devm_kstrdup(dev, "community",
+							      GFP_KERNEL);
```
do we really need to duplicate a constant string?

ujfalusi replied:
I wanted to keep this generic and non Intel's community key specific.

plbossart replied:
that doesn't explain why devm_kstrdup is needed, this duplicates a constant string.

ujfalusi replied:
OK, got it.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -197,6 +197,67 @@ static int sof_machine_check(struct snd_sof_dev *sdev)
+			base_profile->fw_lib_path);
+	else if (base_profile->fw_path_postfix)
+		dev_dbg(dev, "Path postfix appended to default fw_lib path: %s\n",
+			base_profile->fw_path_postfix);
```
clearly not understanding why the fw_path_postfix is used for both the fw_path and fw_lib_path, and what it's supposed to represent.

ujfalusi replied:
because they are firmware files, loaded to firmware and they must have matching postfixes in order to not mix invalid signature.
If the firmware is signed with, lt's say community key then the libraries should also have the same key used.

ujfalusi replied:
This introduces no change in functionality.

plbossart replied:
> because they are firmware files, loaded to firmware and they must have matching postfixes in order to not mix invalid signature. If the firmware is signed with, lt's say community key then the libraries should also have the same key used.

Yes but the community key is the exception to the rule. In generation the base firmware is signed with the Intel production key and the libs with the OEM key.

ujfalusi replied:
> > because they are firmware files, loaded to firmware and they must have matching postfixes in order to not mix invalid signature. If the firmware is signed with, lt's say community key then the libraries should also have the same key used.
> 
> Yes but the community key is the exception to the rule. In generation the base firmware is signed with the Intel production key and the libs with the OEM key.

Oh, I see, hrm, how that should be handled then? We are certainly not taking this into account currently... OEM key is per OEM, right? So we should have Dell, Lenovo, HP, ASUS, etc postfixes for the library?

plbossart replied:
probably yes, the postfix is likely to be modified on a per OEM basis - but IF OEMs decide to release such libraries with some sort of package.
It's also possible that the entire path for libraries is different, why would we put OEM-specific stuff until /lib/firmware/intel ?

ujfalusi replied:
Right, the `fw_path_postfix` is only applied to the default path, if the path is provided via kernel parameter then it is used as it is. I think this makes it a sane rule (no different than what we have currently).

plbossart replied:
we really need a fw_lib_path_postfix.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -290,6 +277,68 @@ static int sof_init_environment(struct snd_sof_dev *sdev)
+	/* init ops, if necessary */
+	ret = validate_sof_ops(sdev);
+	if (ret < 0)
+		return ret;
```
well that clearly will not work with the PR https://github.com/thesofproject/linux/pull/4527 where I am trying to do the exact opposite and initialize the ops in the regular probe, not in the workqueue....

This is becoming complicated.

ujfalusi replied:
If we don't have the `CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE` selected then this is in the probe path, no?
I don't see any issue in the flow.
The only thing is that I will need to rebase or you need to rebase ;)

plbossart replied:
PROBE_WORK_QUEUE needs to remain for the codec parts, so there's really 2 parts of the probe, and we have to initialize stuff in the first part.

ujfalusi replied:
> PROBE_WORK_QUEUE needs to remain for the codec parts, so there's really 2 parts of the probe, and we have to initialize stuff in the first part.

What is going to change regarding to the current flow? The PR works with the workqueue or without it, I cannot imagine a drastic change that would made this not work...

plbossart replied:
the main issue is when the ops are set.

ujfalusi replied:
we should not  have IPC dependent view of the hardware, probing should be IPC neutral.
This series is doing something like that as well, it moves the IPC selection later when we know what the hardware is.

Currently we set the IPC version in stone and then we do a hardware probing but it is not a requirement, the hardware is identical, we just use different set of firmware, topology on it.

plbossart commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -91,37 +198,110 @@ sof_file_profile_for_ipc_type(struct device *dev,
+	 */
+	for (i = 0; i < SOF_IPC_TYPE_COUNT; i++) {
+		if (i == base_profile->ipc_type ||
+		    !(desc->ipc_supported_mask & BIT(i)))
```
so back to my earlier comment, if the sof-bin stuff is not installed for IPC3 we are going to try to setup IPC4?

In addition, if we *know* one IPC is not supported, it should be skipped in this loop, no?

ujfalusi replied:
Exactly.
First we verify the default/selected IPC version.
if a file is missing then we will check each IPC version from 0 and try only the ones:
_not_ the default/selected one (we got here since that was not complete)
it is in the supported list, so we skip IPC versions which is not supported.

In the specific question: if for the platform the IPC3 files are missing, but the platform supports IPC4 then we will try to probe using IPC4, if that is also missing then we fail to probe.

In a reverse and more likely scenario when the platform's default is IPC4 and files missing then we will go and try IPC3.

This all will scale if we have IPC5/6/7/8/9... If the default IPC version is missing files then we will try from the earliest IPC version and try to bring things up.
Hrm, we might want to move backwards from the failed version to pick the newest? It does not matter atm since we have only 2, but we might want to do just that later if needed.

plbossart replied:
> In the specific question: if for the platform the IPC3 files are missing, but the platform supports IPC4 then we will try to probe using IPC4, if that is also missing then we fail to probe.

That's a change of functionality that should be controlled with a Kconfig. For now IPC3 platforms need to work with IPC3 only, otherwise for each case where sof-bin is not properly installed we'll also have a confusing additional error with IPC4 stuff missing, or entire new problems created.

ujfalusi replied:
I see, we have this print in the case you have described:
```
[  115.249648] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw path: community
[  115.249650] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw_lib path: community
[  115.249705] snd_sof:sof_test_firmware_file: sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to open firmware file: intel/sof/community/sof-tgl.ri
[  115.249744] snd_sof:sof_test_firmware_file: sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to open firmware file: intel/sof-ipc4/tgl/community/sof-tgl.ri
[  115.249747] sof-audio-pci-intel-tgl 0000:00:1f.3: SOF firmware and/or topology file not found.
[  115.249749] sof-audio-pci-intel-tgl 0000:00:1f.3: Supported default profiles
[  115.249751] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 0 (Requested):
[  115.249753] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/sof/community/sof-tgl.ri
[  115.249754] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/sof-tplg/sof-hda-generic.tplg
[  115.249756] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 1 (Fallback):
[  115.249758] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/sof-ipc4/tgl/community/sof-tgl.ri
[  115.249760] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/sof-ipc4-tplg/sof-hda-generic.tplg
[  115.249762] sof-audio-pci-intel-tgl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
[  115.249764] sof-audio-pci-intel-tgl 0000:00:1f.3: Optionally it can be manually downloaded from:
[  115.249766] sof-audio-pci-intel-tgl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
```
So to prevent fallback to newer IPC version (is that a fallup?) in case the default cannot be satisfied?
CONFIG_SOF_ALLOW_FALLBACK_TO_NEWER_IPC_VERSION, and it is **n** by default?

plbossart replied:
something like that yes.

plbossart reviewed (commented):
@ujfalusi I tried to look at this PR with new eyes and found a couple of points that I missed earlier... 

The main point is the issue with the ops setup in probe_continue which isn't quite aligned with the need to NOT use the workqueue for i915.

ujfalusi commented:
> @ujfalusi I tried to look at this PR with new eyes and found a couple of points that I missed earlier...

Thank you for  a fresh look!

> The main point is the issue with the ops setup in probe_continue which isn't quite aligned with the need to NOT use the workqueue for i915.

I don't think this is an issue since if we don't have the `CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE` selected, then all is executed in the probe at the end.
It works in a similar way as the current code works, we just move things around to be able to determine of the IPC version can be used in the system.

ujfalusi commented:
Changes since v5:
- remove the "IS_ERR_OR_NULL(plat_data->pdev_mach)" check before calling platform_device_unregister()
- New Kconfig option restrict the fallback to older IPC versions by default

plbossart commented:
@ujfalusi @kv2019i I don't have any sustained objections to the introduction of profiles, just that this needs to happen after the changes needed for i915/xe since I do think there is a conflict on when the ops are set.

ujfalusi commented:
> @ujfalusi @kv2019i I don't have any sustained objections to the introduction of profiles, just that this needs to happen after the changes needed for i915/xe since I do think there is a conflict on when the ops are set.

OK, fair enough, I will do the rebase when it is time and let's see then how things play out. Fwiw, it did not looked doable with the current code either before I started to focus on this.

ujfalusi commented:
Changes since v6:
- rebased on topic/sof-dev to fix conflict

ujfalusi commented:
Changes since v7:
- patches re-arranged and squashed when it made sense
 -  the IPC type fallback and the Kconfig patch to limit the forward fallback
 - the LNL is no longer separate for the manual managed ipc4_data

plbossart commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -0,0 +1,127 @@
+			out_profile->fw_lib_path = devm_kasprintf(dev,
+							GFP_KERNEL, "%s/%s",
+							desc->default_lib_path[ipc_type],
+							base_profile->fw_path_postfix);
```
again I think it's not right to conflate fw and lib signatures. we know it doesn't fly for Windows OEMs.

plbossart reviewed (changes requested):
Still two points that are unclear

a) the support for different signature keys for fw and libs does not seem to exist.
b) there's still no clarity on conflicts with the i915 ops changes (which doesn't seem to progress either).

No other objections.

ujfalusi commented:
> Still two points that are unclear
> 
> a) the support for different signature keys for fw and libs does not seem to exist.

@plbossart, I really don't know what is expected to be done regarding to this. The PR introduces no changes regarding to this. If it needs changes then it has to be a separate PR.

> b) there's still no clarity on conflicts with the i915 ops changes (which doesn't seem to progress either).

I have a local rebase now on top of the i915 deferred probe handling, all is good, I will update the PR.

> No other objections.

ujfalusi commented:
Changes since v8:
- rebased on topic/sof-dev which now includes the i915 deferred probe handling
- `sof_init_sof_ops()` introduced because we need ops to be setup before calling `snd_sof_probe_early()` - this is the 'big' change needed by the deferred probe handling.

plbossart commented:
> > Still two points that are unclear
> > a) the support for different signature keys for fw and libs does not seem to exist.
> 
> @plbossart, I really don't know what is expected to be done regarding to this. The PR introduces no changes regarding to this. If it needs changes then it has to be a separate PR.

Well, this PR is described as preparing the future, but there's one shortcoming that was identified. If we do the support for libraries later, it adds a second dependency between firmware and kernel, so why not do the right thing now?

ujfalusi commented:
> > > Still two points that are unclear
> > > a) the support for different signature keys for fw and libs does not seem to exist.
> > 
> > 
> > @plbossart, I really don't know what is expected to be done regarding to this. The PR introduces no changes regarding to this. If it needs changes then it has to be a separate PR.
> 
> Well, this PR is described as preparing the future, but there's one shortcoming that was identified. If we do the support for libraries later, it adds a second dependency between firmware and kernel, so why not do the right thing now?

So, we should deal with the libraries first?
What we have atm is a default location for the libraries, which is `lib_path` (+`community` in case of community key platform).
We also have a lib path override option which can be used to specify the path (used  as it is, not extended with `community`).

I don't see how we could facilitate vendor key use for Lenovo, Dell, HP, ASUS, etc in a default path. We would need a 'huge' match table to set different vendor directories and in some cases different directories per vendors depending on keys?

This PR is just moving the path handling into core to a common location and the last patch adds support for moving back (and forth) in IPC version if files for the selected version were missing, but it only deals with default paths, custom paths are used as provided.

The followup series does have the definition of the default paths moved to core, so there we really should think how we can do this.

I was thinking of tplg token to tell the lib location, but the same tplg can be used on different vendor's device, so that is not good.

ujfalusi commented:
@plbossart, I do want to keep this and the next PR 'simple' by not introducing ABI (as in file location) changes if possible.

Having said that and after some time to think about the libraries, I think what might make sense is:

-- default library path --
The libraries we, as SOF project release, signed with the same key as the firmware.
either
`<vendor>/sof-ipc4-lib/<platform></fw_path_postfix>`
or
`<vendor>/sof-ipc4/<platform></fw_path_postfix>` iow, along with the base firmware file

-- customer library path --
Libraries provided by customers (Dell, Lenovo, HP, Google (?), etc)
either
`<vendor>/sof-ipc4-lib/<platform>/<customer>`
or
`<vendor>/sof-ipc4/<platform>/<customer>`

I have no idea how the customer path can be provided, but it would be used together with the 'core' lib path:
first we would look for the lib in **customer** path, if the lib is not there then we would look under the **SOF** path.

Deployment might be an issue for distributions due to the shared path, but that can be tuned?

I kind of like the idea to have the default path for libraries pointing to the firmware location (less hassle in code and documentation).

We can move the discussion to some better place (sof-docs?) and settle  the final places.

FWIW, this PR and the next one pending will gave us quite a bit of flexibility on moving things around - given that the kernel have those patches - since we can declare different 'default' layouts for IPC versions to cover the past, present and future.

Solely on this PR: this is not introducing anything new in paths or how they are provided, and yes, at the moment we use the same path 'patching' for the fw path and the fw library path:
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/sof-pci-dev.c#L266-L307

plbossart commented:
My only objection is that "fw_path_postfix" needs to be used for firmware only. We need a separate variable for libraries. How "fw_libs_path_postfix" is set for customers is something we debate in the future. The only thing we need to consider NOW is that the hardware allows for two separate keys, so we cannot keep assuming that we have a common key between firmware and libs.

For example, we could test on Up Extreme with the firmware signed with the production key and the libraries signed with the community key. this should work (famous last words I know).

marc-hb commented:
> ... with the firmware signed with the production key and the libraries signed with the community key. this should work (famous last words I know).

I haven't looked at all the details, only followed these discussions from a distance. However I'm pretty sure one of the main purposes of loadable libraries is to allow them to be signed with a variety of keys. Not as some future improvement but from the moment they're supported.

Does that mean libraries should be located in some key-specific directories? Does this require some "huge" vendor table? No idea but about time this gets designed.

> We can move the discussion to some better place (sof-docs?) and settle the final places.

+1. Prototype code is super useful, among others because it helps discover gaps. Burying major design choices at the bottom of very long PRs is not so good.

-- 
Marc "sof-bin" Herbert

ujfalusi commented:
Changes since v9:
- Add `fw_lib_path_postfix` to `struct sof_loadable_file_profile` to prepare the code to support separate path postfix for base firmware and libraries. The `fw_lib_path_postfix` is set to the same string for now to not introduce any change regarding to the paths used currently.

keqiaozhang commented:
SOFCI TEST

plbossart commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -94,37 +201,122 @@ sof_file_profile_for_ipc_type(struct device *dev,
 
-static void sof_print_profile_info(struct device *dev,
+static void
+sof_missing_firmware_notification(struct snd_sof_dev *sdev,
```
nit-pick: usually "firmware notification" means some sort of firmware-generated IPC is received by the host driver. This is not what this function does. I think this is a user notification, but that's confusing.

ujfalusi replied:
OK, I will find a better name for the function.

marc-hb replied:
`sof_log_missing_firmware_file()`

plbossart reviewed (dismissed):
LGTM - except for one confusing function name, see below @ujfalusi

ujfalusi commented:
Changes since v10:
- rename sof_missing_firmware_notification() to sof_print_missing_firmware_info() - to be inline with the 'normal' print function name.

kv2019i commented on sound/soc/sof/core.c:
```diff
@@ -204,6 +204,67 @@ static int sof_machine_check(struct snd_sof_dev *sdev)
+			base_profile->fw_path);
+	else if (base_profile->fw_path_postfix)
+		dev_dbg(dev, "Path postfix appended to default fw path: %s\n",
+			base_profile->fw_path_postfix);
```
I guess this is intentionally either-or?

ujfalusi replied:
yes it is intentional, if the path is provided as kernel parameter then we don't change it, user must know what he/she is doing.

kv2019i commented on sound/soc/sof/intel/apl.c:
```diff
@@ -55,7 +55,7 @@ int sof_apl_ops_init(struct snd_sof_dev *sdev)
 		struct sof_ipc4_fw_data *ipc4_data;
 
-		sdev->private = devm_kzalloc(sdev->dev, sizeof(*ipc4_data), GFP_KERNEL);
+		sdev->private = kzalloc(sizeof(*ipc4_data), GFP_KERNEL);
```
Hmm, should we have a kfree() in hda_ops_free() then?

ujfalusi replied:
see the change in hda-dai.c in `hda_ops_free()` ;)

kv2019i replied:
Ack @ujfalusi not sure how I missed that , but indeed it's there.

kv2019i reviewed (commented):
Looks very nice @ujfalusi ! Can you check one mem alloc question inline?
